pkg/ccl/changefeedccl/alter_changefeed_test.go

```diff
@@ -1769,3 +1882,271 @@ func TestAlterChangefeedAccessControl(t *testing.T) {
+				_, err := fmt.Fprintf(&alterStmtBuilder, format, args...)
+				require.NoError(t, err)
+			}
+			write(`ALTER CHANGEFEED %d`, feed.JobID())
+
+			// We get the set of tables to add/drop first to ensure we are
+			// selecting without replacement.
+			numAdds := rand.Intn(len(nonWatchedTables) + 1)
+			numDrops := rand.Intn(len(watchedTables))
+			if numAdds == 0 && numDrops == 0 {
+				t.Logf("skipping alter #%d", i+1)
+				continue
+			}
+			adds := getNFromSet(nonWatchedTables, numAdds)
+			drops := getNFromSet(watchedTables, numDrops)
+
+			var expectedRows []string
+			for len(adds) > 0 || len(drops) > 0 {
+				// Randomize the order of adds and drops.
+				switch add := len(adds) > 0 && (len(drops) == 0 || rnd.Intn(2) == 0); add {
+				case true: // Add a target.
+					addTarget := getOneFromSet(adds)
+					delete(adds, addTarget)
+					delete(nonWatchedTables, addTarget)
+					watchedTables[addTarget] = struct{}{}
+
+					write(` ADD %s`, addTarget)
+
+					switch rnd.Intn(4) {
+					case 0:
+						write(` WITH initial_scan='yes'`)
+						expectedRows = append(expectedRows, makeInitialScanRows([]string{addTarget}, hw)...)
+					case 1:
+						write(` WITH initial_scan='only'`)
+						// We don't do an initial scan because the original
+						// changefeed did not have initial_scan='only'.
+					case 2:
+						write(` WITH initial_scan='no'`)
+					case 3:
+						// The default option is initial_scan='no'.
+					}
+					expectedRows = append(expectedRows,
+						insertRowsIntoTable(addTarget, 2 /* numRows */)...)
+				case false: // Drop a target.
+					dropTarget := getOneFromSet(drops)
+					delete(drops, dropTarget)
+					delete(watchedTables, dropTarget)
+					nonWatchedTables[dropTarget] = struct{}{}
+
+					write(` DROP %s`, dropTarget)
+
+					// Insert some more rows into the table. They should NOT be emitted by the changefeed.
+					insertRowsIntoTable(dropTarget, 3 /* numRows */)
+				}
+			}
+			require.Empty(t, adds)
+			require.Empty(t, drops)
+
+			alterStmt := alterStmtBuilder.String()
+			t.Log(alterStmt)
+			sqlDB.Exec(t, alterStmt)
+
+			require.NoError(t, feed.Resume())
+
+			// Wait for highwater to advance past the current time so that
+			// we're sure no more rows are expected.
+			var tsStr string
+			sqlDB.QueryRow(t, `SELECT cluster_logical_timestamp()`).Scan(&tsStr)
+			ts := parseTimeToHLC(t, tsStr)
+			require.NoError(t, feed.WaitForHighWaterMark(ts))
+
+			assertPayloads(t, testFeed, expectedRows)
+		}
+	}
+
+	cdcTest(t, testFn, feedTestEnterpriseSinks, feedTestNoExternalConnection)
+}
+
+// setDifference returns a new set that is s - t.
+func setDifference[K comparable](s map[K]struct{}, t map[K]struct{}) map[K]struct{} {
+	difference := make(map[K]struct{})
+	for e := range s {
+		if _, ok := t[e]; !ok {
+			difference[e] = struct{}{}
+		}
+	}
+	return difference
+}
+
+// getOneFromSet returns a random element from s.
+func getOneFromSet[K comparable](s map[K]struct{}) K {
+	for e := range s {
```

Comment:
nit: do you want to use the pseudorandom generator for this too? ditto below

Reply:
Good call, I made some changes so the test should be reproducible with the same seed now